Mathlib/NumberTheory/Cyclotomic/Rat.lean

```diff
@@ -331,16 +332,104 @@ theorem zeta_sub_one_prime [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
 /-- `ζ - 1` is prime if `ζ` is a primitive `p`-th root of unity. -/
 theorem zeta_sub_one_prime' [h : IsCyclotomicExtension {p} ℚ K] (hζ : IsPrimitiveRoot ζ p) :
     Prime ((hζ.toInteger - 1)) := by
-  convert zeta_sub_one_prime (k := 0) (by simpa)
-  simpa
+  convert zeta_sub_one_prime (k := 0) (by simpa only [zero_add, pow_one])
+  simpa only [zero_add, pow_one]
 
 theorem subOneIntegralPowerBasis_gen_prime [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
     (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) :
-    Prime hζ.subOneIntegralPowerBasis.gen := by simpa using hζ.zeta_sub_one_prime
+    Prime hζ.subOneIntegralPowerBasis.gen := by
+  simpa only [subOneIntegralPowerBasis_gen] using hζ.zeta_sub_one_prime
 
 theorem subOneIntegralPowerBasis'_gen_prime [IsCyclotomicExtension {p} ℚ K]
     (hζ : IsPrimitiveRoot ζ ↑p) :
-    Prime hζ.subOneIntegralPowerBasis'.gen := by simpa using hζ.zeta_sub_one_prime'
+    Prime hζ.subOneIntegralPowerBasis'.gen := by
+  simpa only [subOneIntegralPowerBasis'_gen] using hζ.zeta_sub_one_prime'
+
+open nonZeroDivisors in
+/-- The norm, relative to `ℤ`, of `ζ ^ p ^ s - 1` in a `p ^ (k + 1)`-th cyclotomic extension of `ℚ`
+is p ^ p ^ s` if `s ≤ k` and `p ^ (k - s + 1) ≠ 2`. -/
+lemma norm_toInteger_pow_sub_one_of_prime_pow_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) {s : ℕ} (hs : s ≤ k) (htwo : p ^ (k - s + 1) ≠ 2) :
+    Algebra.norm ℤ (hζ.toInteger ^ (p : ℕ) ^ s - 1) = p ^ (p : ℕ) ^ s := by
+  apply (algebraMap ℤ ℚ).injective_int
+  have : NumberField K := IsCyclotomicExtension.numberField {p ^ (k + 1)} ℚ K
+  have : algebraMap (𝓞 K) K (hζ.toInteger ^ (p : ℕ) ^ s - 1) = ζ ^ (p : ℕ) ^ s - 1 := by
+    simp only [map_sub, map_pow, map_one, sub_left_inj, RingOfIntegers.map_mk]
+  rw [← Algebra.norm_localization (Sₘ := K) ℤ ℤ⁰, this,
+    hζ.norm_pow_sub_one_of_prime_pow_ne_two
+      (cyclotomic.irreducible_rat (by simp only [PNat.pow_coe, gt_iff_lt, PNat.pos, pow_pos]))
+      hs htwo]
+  simp only [algebraMap_int_eq, map_pow, map_natCast]
+
+open nonZeroDivisors in
+/-- The norm, relative to `ℤ`, of `ζ ^ 2 ^ k - 1` in a `2 ^ (k + 1)`-th cyclotomic extension of `ℚ`
+is `(-2) ^ 2 ^ k`. -/
+lemma norm_toInteger_pow_sub_one_of_two [IsCyclotomicExtension {2 ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑((2 : ℕ+) ^ (k + 1))) :
+    Algebra.norm ℤ (hζ.toInteger ^ 2 ^ k - 1) = (-2) ^ (2 : ℕ) ^ k := by
+  apply (algebraMap ℤ ℚ).injective_int
+  have : NumberField K := IsCyclotomicExtension.numberField {2 ^ (k + 1)} ℚ K
+  have : algebraMap (𝓞 K) K (hζ.toInteger ^ 2 ^ k - 1) = ζ ^ (2 : ℕ) ^ k - 1 := by
+    simp only [map_sub, map_pow, map_one, sub_left_inj, RingOfIntegers.map_mk]
+  rw [← Algebra.norm_localization (Sₘ := K) ℤ ℤ⁰, this,
+    hζ.norm_pow_sub_one_two (cyclotomic.irreducible_rat (pow_pos (by decide) _))]
+  simp only [algebraMap_int_eq, Int.reduceNeg, map_pow, map_neg, map_ofNat]
```

Comment:
Maybe we should extract the following proof step into its own lemma? The only thing that I would like to add is somehow turning `hζ` into `ζ` in the new lemma, but I don't think there's a good way to do so. (I think the statement is also generalizable quite a bit, didn't put much effort in.)
```suggestion
open nonZeroDivisors in
/-- The norm of `hζ : 𝓞 K` in `ℤ` can be computed in `ℚ`. -/
lemma norm_toInteger_eq_iff [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) {f : 𝓞 K → 𝓞 K} {y : ℤ} :
    Algebra.norm ℤ (f hζ.toInteger) = y ↔
      (Algebra.norm ℚ) ((algebraMap (𝓞 K) K) (f hζ.toInteger)) = ↑y := by
  have : NumberField K := IsCyclotomicExtension.numberField {p ^ (k + 1)} ℚ K
  rw [← (algebraMap ℤ ℚ).injective_int.eq_iff, ← Algebra.norm_localization (Sₘ := K) ℤ ℤ⁰,
      eq_intCast]

/-- The norm, relative to `ℤ`, of `ζ ^ p ^ s - 1` in a `p ^ (k + 1)`-th cyclotomic extension of `ℚ`
is p ^ p ^ s` if `s ≤ k` and `p ^ (k - s + 1) ≠ 2`. -/
lemma norm_toInteger_pow_sub_one_of_prime_pow_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) {s : ℕ} (hs : s ≤ k) (htwo : p ^ (k - s + 1) ≠ 2) :
    Algebra.norm ℤ (hζ.toInteger ^ (p : ℕ) ^ s - 1) = p ^ (p : ℕ) ^ s := by
  rw [norm_toInteger_eq_iff (f := fun x => x ^ (p : ℕ) ^ s - 1)]
  simp [hζ.norm_pow_sub_one_of_prime_pow_ne_two
          (cyclotomic.irreducible_rat (by simp only [PNat.pow_coe, gt_iff_lt, PNat.pos, pow_pos]))
          hs htwo]

/-- The norm, relative to `ℤ`, of `ζ ^ 2 ^ k - 1` in a `2 ^ (k + 1)`-th cyclotomic extension of `ℚ`
is `(-2) ^ 2 ^ k`. -/
lemma norm_toInteger_pow_sub_one_of_two [IsCyclotomicExtension {2 ^ (k + 1)} ℚ K]
    (hζ : IsPrimitiveRoot ζ ↑((2 : ℕ+) ^ (k + 1))) :
    Algebra.norm ℤ (hζ.toInteger ^ 2 ^ k - 1) = (-2) ^ (2 : ℕ) ^ k := by
  rw [norm_toInteger_eq_iff (f := fun x => x ^ 2 ^ k - 1)]
  simp [hζ.norm_pow_sub_one_two (cyclotomic.irreducible_rat (pow_pos (by decide) _))]
```

Reply:
`norm_toInteger_eq_iff` looks a reasonable idea, but it has nothing to do with cyclotomic fields. I will move it where it should be probably tomorrow, thanks!